rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,379 @@
+  x = encoder([input_ids, input_mask, label_ids])['pooled_output']
+  cls_output, sequence_output = output['pooled_output'], outputs['sequence_output']
+  masked_lm = keras_nlp.layers.MaskedLM(embedding_table=encoder.get_embedding_table())
+  lm_output = masked_lm(sequence_output, masked_positions=masked_lm_positions)
+  cls_output = tf.keras.layers.Dense(units=num_classes, activation='softmax')(cls_output)
+  model = tf.keras.Model(inputs=[input_ids, input_mask, label_ids, masked_lm_positions],
+                         outputs={'lm_output': masked_lm, 'cls_output': cls_output})
+
+model.compile('adam', {'lm_output': 'sparse_categorical_crossentropy', 'cls_output': 'sparse_categorical_crossentropy'})
+model.fit(train_ds, epochs=100)
+```
+
+### Other encoder-based networks
+
+`BertEncoder` is the first encoder network we propose in this doc. However other encoder networks can be easily
+built on top of the `TransformerEncoderBlock` layer. For example, for a transformer encoder sharing mechanism
+with [ALBERT](https://arxiv.org/pdf/1909.11942.pdf), this can be achieved by:
+
+```python
+word_ids = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_word_ids')
+mask = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_mask')
+type_ids = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_type_ids')
+embedding_layer = keras_nlp.layers.OnDeviceEmbedding(vocab_size, embedding_width)
+word_embeddings = embedding_layer(word_ids)
+position_embedding_layer = keras_nlp.layers.PositionEmbedding(max_sequence_length)
+position_embeddings = position_embedding_layer(word_embeddings)
+type_embeddings = keras_nlp.layers.OnDeviceEmbedding(
+  vocab_size=type_vocab_size, embedding_width=embedding_width, use_one_hot=True)(type_ids)
+embeddings = tf.keras.layers.Add()([word_embeddings, position_embeddings, type_embeddings])
+embeddings = tf.keras.layers.LayerNormalization(axis=-1)(embeddings)
+embeddings = tf.keras.layers.Dropout(rate=dropout_rate)(embeddings)
+embeddings = tf.keras.layers.experimental.EinsumDense(
+  '...x,xy->...y', output_shape=hidden_size, bias_axes='y')(embeddings)
+data = emnbeddings
+attention_mask = layers.SelfAttentionMask()([data, mask])
+shared_layer = keras_nlp.layers.TransformerEncoderBlock(num_attention_heads, inner_dim)
+for _ in range(num_layers):
+  data = shared_layer([data, attention_mask])
+first_token_tensor = tf.keras.layers.Lambda(lambda x: tf.squeeze(x[:, 0:1, :], axis=1))(data)
+cls_output = tf.keras.layers.Dense(units=hidden_size, activation='tanh')(first_token_tensor)
+outputs = dict(sequence_output=data, pooled_output=cls_output)
+model = tf.keras.Model(inputs=[word_ids, mask, type_ids], outputs=outputs)
+```
+
+## Detailed Design
+
+### Layers -- TransformerEncoderBlock
+
+This layer encapsulates a single layer of Transformer Encoder.
+
+```python
+class TransformerEncoderBlock(tf.keras.layers.Layer):
```

Comment:
Why not `TransformerEncoder`?

Reply:
I tried to rename it to be `TransformerEncoder`, but was pointed out that this is a single transformer layer and given Bert usually need to use 12 of them.
Second thought we should name this to be `TransformerEncoder` though. Updated.

Reply:
@saberkun FYI.